File: java/collection.joookiwi/src/main/java/joookiwi/collection/java/method/HasDuplicate.java
package joookiwi.collection.java.method;

import joookiwi.collection.java.CollectionHolder;
import joookiwi.collection.java.MinimalistCollectionHolder;
import joookiwi.collection.java.annotation.ExtensionFunction;
import joookiwi.collection.java.exception.ImpossibleConstructionException;
import org.jetbrains.annotations.Contract;
import org.jetbrains.annotations.NotNullByDefault;
import org.jetbrains.annotations.Nullable;
import org.jetbrains.annotations.Unmodifiable;

import static joookiwi.collection.java.CommonContracts.ALWAYS_FAIL_0;
import static joookiwi.collection.java.CommonContracts.IF_1ST_NULL_THEN_FALSE_1;

@NotNullByDefault
public final class HasDuplicate
    extends Utility {

    @Contract(ALWAYS_FAIL_0)
    private HasDuplicate() { throw new ImpossibleConstructionException("The utility class “HasDuplicate” cannot be constructed.", HasDuplicate.class); }

    //#region -------------------- Facade methods --------------------

    /// The `collection` has at least one duplicate value
    ///
    /// @param collection The [nullable][Nullable] [collection][MinimalistCollectionHolder]
    /// @return **true** only if one element is equal (== or [equals][#equals]) to another one
    @ExtensionFunction
    @Contract(IF_1ST_NULL_THEN_FALSE_1)
    public static <T extends @Nullable Object> boolean hasDuplicate(final @Nullable MinimalistCollectionHolder<? extends T> collection) {
        if (collection == null)
            return false;

        final var size = collection.size();
        if (size == 0)
            return false;
        return __hasDuplicate(collection, size);
    }

    /// The `collection` has at least one duplicate value
    ///
    /// @param collection The [nullable][Nullable] [collection][CollectionHolder]
    /// @return **true** only if one element is equal (== or [equals][#equals]) to another one
    @ExtensionFunction
    @Contract(IF_1ST_NULL_THEN_FALSE_1)
    public static <T extends @Nullable Object> boolean hasDuplicate(final @Nullable CollectionHolder<? extends T> collection) {
        if (collection == null)
            return false;
        if (collection.isEmpty())
            return false;
        return __hasDuplicate(collection, collection.size());
    }

    /// The `collection` has at least one duplicate value
    ///
    /// @param collection The [nullable][Nullable] collection
    /// @return **true** only if one element is equal (== or [equals][#equals]) to another one
    @ExtensionFunction
    @Contract(IF_1ST_NULL_THEN_FALSE_1)
    public static <T extends @Nullable Object> boolean hasDuplicate(final T @Nullable @Unmodifiable [] collection) {
        if (collection == null)
            return false;

        final var size = collection.length;
        if (size == 0)
            return false;
        return __hasDuplicate(collection, size);
    }

    //#endregion -------------------- Facade methods --------------------
    //#region -------------------- Loop methods --------------------

    private static <T extends @Nullable Object> boolean __hasDuplicate(final MinimalistCollectionHolder<? extends T> collection,
                                                                       final int size) {
        @SuppressWarnings("unchecked cast") final var temporaryArray = (T[]) new Object[size];
        temporaryArray[0] = collection.get(0);
        var amountOfItemAdded = 1;
        var index = 0;
        while (++index < size) {
            final var value = collection.get(index);
            var index2 = -1;

            if (value == null) {
                while (++index2 < amountOfItemAdded)
                    if (temporaryArray[index2] == null)
                        return true;
                temporaryArray[amountOfItemAdded++] = null;
                continue;
            }

            while (++index2 < amountOfItemAdded) {
                final var comparedValue = temporaryArray[index2];
                if (comparedValue == value || value.equals(comparedValue))
                    return true;
            }
            temporaryArray[amountOfItemAdded++] = value;
        }

        return amountOfItemAdded != size;
    }

    private static <T extends @Nullable Object> boolean __hasDuplicate(final T @Unmodifiable [] collection,
                                                                       final int size) {
        @SuppressWarnings("unchecked cast") final var temporaryArray = (T[]) new Object[size];
        temporaryArray[0] = collection[0];
        var amountOfItemAdded = 1;
        var index = 0;
        while (++index < size) {
            final var value = collection[index];
            var index2 = -1;

            if (value == null) {
                while (++index2 < amountOfItemAdded)
                    if (temporaryArray[index2] == null)
                        return true;
                temporaryArray[amountOfItemAdded++] = null;
                continue;
            }

            while (++index2 < amountOfItemAdded) {
                final var comparedValue = temporaryArray[index2];
                if (comparedValue == value || value.equals(comparedValue))
                    return true;
            }
            temporaryArray[amountOfItemAdded++] = value;
        }

        return amountOfItemAdded != size;
    }

    //#endregion -------------------- Loop methods --------------------

}

File: javascript/src/method/hasDuplicate.ts
//··························································
// Copyright (c) 2023-2026. Jonathan Bédard ~ JóôòKiwi
//
// This project is free to use.
// All the right is reserved to the author of this project.
// My projects:
//  - https://github.com/joooKiwi/type
//  - https://github.com/joooKiwi/lazy
//  - https://github.com/joooKiwi/collection
//  - https://github.com/joooKiwi/enumeration
//··························································

import type {Nullable} from "@joookiwi/type"

import type {CollectionHolder}           from "../CollectionHolder"
import type {MinimalistCollectionHolder} from "../MinimalistCollectionHolder"

import {isArray}                       from "./isArray"
import {isArrayByStructure}            from "./isArrayByStructure"
import {isCollectionHolder}            from "./isCollectionHolder"
import {isCollectionHolderByStructure} from "./isCollectionHolderByStructure"
import {isMinimalistCollectionHolder}  from "./isMinimalistCollectionHolder"

//#region -------------------- Facade method --------------------

/**
 * The {@link collection} has at least one duplicate value
 *
 * @param collection The {@link Nullable nullable} collection ({@link MinimalistCollectionHolder}, {@link CollectionHolder} or {@link ReadonlyArray Array})
 * @return {boolean} <b>true</b> only if one element is equal (===) to another one
 * @extensionFunction
 */
export function hasDuplicate<const T, >(collection: Nullable<| MinimalistCollectionHolder<T> | CollectionHolder<T> | readonly T[]>,): boolean {
    if (collection == null)
        return false
    if (isCollectionHolder(collection,))
        return hasDuplicateByCollectionHolder(collection,)
    if (isArray(collection,))
        return hasDuplicateByArray(collection,)
    if (isMinimalistCollectionHolder(collection,))
        return hasDuplicateByMinimalistCollectionHolder(collection,)

    if (isCollectionHolderByStructure<T>(collection,))
        return hasDuplicateByCollectionHolder(collection,)
    if (isArrayByStructure<T>(collection,))
        return hasDuplicateByArray(collection,)
    return hasDuplicateByMinimalistCollectionHolder(collection,)
}


/**
 * The {@link collection} has at least one duplicate value
 *
 * @param collection The nullable collection
 * @return {boolean} <b>true</b> only if one element is equal (===) to another one
 * @extensionFunction
 */
export function hasDuplicateByMinimalistCollectionHolder<const T, >(collection: Nullable<MinimalistCollectionHolder<T>>,): boolean {
    if (collection == null)
        return false

    const size = collection.size
    if (size == 0)
        return false
    if (size == 1)
        return false
    return __validate(collection, size,)
}

/**
 * The {@link collection} has at least one duplicate value
 *
 * @param collection The nullable collection
 * @return {boolean} <b>true</b> only if one element is equal (===) to another one
 * @extensionFunction
 */
export function hasDuplicateByCollectionHolder<const T, >(collection: Nullable<CollectionHolder<T>>,): boolean {
    if (collection == null)
        return false
    if (collection.isEmpty)
        return false

    const size = collection.size
    if (size == 1)
        return false
    return __validate(collection, size,)
}

/**
 * The {@link collection} has at least one duplicate value
 *
 * @param collection The nullable collection
 * @return {boolean} <b>true</b> only if one element is equal (===) to another one
 * @extensionFunction
 */
export function hasDuplicateByArray<const T, >(collection: Nullable<readonly T[]>,): boolean {
    if (collection == null)
        return false

    const size = collection.length
    if (size == 0)
        return false
    if (size == 1)
        return false
    return __validateByArray(collection, size,)
}

//#endregion -------------------- Facade method --------------------
//#region -------------------- Loop methods --------------------

function __validate(collection: MinimalistCollectionHolder, size: number,) {
    const temporaryArray = new Array(size,)
    temporaryArray[0] = collection.get(0,)
    let amountOfItemAdded = 1
    let index = 0
    while (++index < size) {
        const value = collection.get(index,)
        let index2 = -1
        while (++index2 < amountOfItemAdded)
            if (temporaryArray[index2] === value)
                return true
        temporaryArray[amountOfItemAdded++] = value
    }

    return false
}

function __validateByArray(collection: readonly unknown[], size: number,) {
    const temporaryArray = new Array(size,)
    temporaryArray[0] = collection[0]
    let amountOfItemAdded = 1
    let index = 0
    while (++index < size) {
        const value = collection[index]
        let index2 = -1
        while (++index2 < amountOfItemAdded)
            if (temporaryArray[index2] === value)
                return true
        temporaryArray[amountOfItemAdded++] = value
    }

    return false
}

//#endregion -------------------- Loop methods --------------------
